kubernetes/google-kubernetes-engine/bookie.stateful.yaml

```diff
@@ -0,0 +1,170 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+## Credits to Franck Cuny : https://github.com/fcuny/distributedlog-on-k8s/blob/master/bookkeeper.statefulset.yaml
+
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: ssd
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-ssd
+
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: bookie-config
+data:
+  PULSAR_MEM: "\" -Xms4g -Xmx4g -XX:MaxDirectMemorySize=4g\""
+  PULSAR_GC: "\" -XX:+UseG1GC \""
+  dbStorage_writeCacheMaxSizeMb: "1024"
+  dbStorage_readAheadCacheMaxSizeMb: "1024"
+  zkServers: zk-0.zookeeper,zk-1.zookeeper,zk-2.zookeeper
+  statsProviderClass: org.apache.bookkeeper.stats.PrometheusMetricsProvider
+---
+
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: bookie
+  labels:
+    app: pulsar
+    component: bookkeeper
+spec:
+  serviceName: "bookkeeper"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: pulsar
+        component: bookkeeper
+        cluster: pulsar-gke
+      annotations:
+        pod.alpha.kubernetes.io/initialized: "true"
+        prometheus.io/scrape: "true"
+        prometheus.io/port: "8000"
+    spec:
+      terminationGracePeriodSeconds: 0
+      containers:
+        - name: bookie
+          image: apachepulsar/pulsar:latest
+          command: ["sh", "-c"]
+          args:
+            - >
+              bin/apply-config-from-env.py conf/bookkeeper.conf &&
+              bin/apply-config-from-env.py conf/pulsar_env.sh &&
+              bin/pulsar bookie
+          ports:
+            - containerPort: 3181
+              hostPort: 3181
+              name: client
+          envFrom:
+            - configMapRef:
+                name: bookie-config
+          env:
+            - name: advertisedAddress
+              valueFrom:
+                fieldRef:
+                  fieldPath: status.hostIP
+          volumeMounts:
+            - name: journal-disk
+              mountPath: /pulsar/data/bookkeeper/journal
+            - name: ledgers-disk
+              mountPath: /pulsar/data/bookkeeper/ledgers
+      initContainers:
+        # The first time, initialize BK metadata in zookeeper
+        # Otherwise ignore error if it's already there
+        - name: bookie-metaformat
+          image: apachepulsar/pulsar:latest
+          command: ["sh", "-c"]
+          args:
+            - >
+              bin/apply-config-from-env.py conf/bookkeeper.conf &&
+              bin/bookkeeper shell metaformat --nonInteractive || true;
+          envFrom:
+            - configMapRef:
+                name: bookie-config
+
+      volumes:
+          # Mount local disks
+        - name: journal-disk
+          hostPath:
```

Comment:
If we're using StatefulSet, we cannot just mount a volume on the local disk, because the Pod might be rescheduled somewhere else. I think we should just stick with a network attached PersistentVolume, with SSD for journal and HDD for ledgers storage.